### ISSUE TITLE
Font variants

### ISSUE THREAD

**Issue by henryso** [gregorio-project/gregorio#301]
Gregorio currently supports in-font variants for punctum cavum and out-of-font variants for clefs and flats.  I (and some others, apparently) would like a different shape for the stropha in Greciliae.  We should probably implement some sort of variant system.

* This can be as simple as defining the additional glyphs and letting users redefine the ones they wish to override: `\let\grecpStropha=\grecpStrophaVariantB`
* Or perhaps the same with the assistance of a macro and some glyph naming system: `\greusealternative{Stropha}{VariantB}`
* Or something like what we have for `\greusehufnagelstyle`
* Or some other concoction of our imagination...

Thoughts?

**Comment by eroux** [comment]
Maybe we should have something more generic in mind... Here are my 2c:

- there should be a common way of addressing alternative glyphs, the usual way in modern fonts are `xxx.alt1`, `xxx.alt2`
- it might be relevant to set the font for an alternate, this would resolve the case of greextra, but would also open new cases like "I want to use greciliae but with the stropha of Gregoria" or something like that. It would be something like `\greAlternative{stropha}{greciliae.stropha.alt1}`. The difficulty is that horizontal episemus might not have the appropriate length... I don't really know how to deal with that...

**Comment by henryso** [comment]
I think we can punt on the horizontal episemus issue.  If the user really wants to use that alternative from a different font, then they have to live with what they get.

**Comment by eroux** [comment]
I think if we have some conventions, that would be possible... Furthermore, having a hepisemus by variant is mandatory, for instance if we set the stropha to the new one, the hepisemus might shrink a little... So maybe having hepisemus with the name `glyph.hep` or so would make it possible, when changing a glyph for an alternate, to also change the hepisemus for the alternate? What do you think?

**Comment by henryso** [comment]
There are 19 hepisemus glyphs and some of them are re-used for multiple sets of glyphs.  If we just allow them to be redefined in the same way other glyphs are redefined, I think that's probably enough, rather than doing something automatically based on glyph name.

**Comment by eroux** [comment]
Ok, with a bit of documentation, that's probably enough... What about taking an alternative glyph in a different font?

**Comment by rpspringuel** [comment]
I would start with a simple system that allows manual substitution of one glyph for another.  Ideally the new glyph should be able to come from a list of predefined alternatives, one of the other fonts, or even a custom font the user might have on their own system.  While the underlying process may use the TeX primitive `\let` (and those substituting from a custom font might have to use this mechanism), I would favor having a user level macro (or two) which makes interacting with the underlying structure easier when dealing with the predefined alternatives or glyphs from another of "our" fonts.

I would also say that the user can be left responsible for ironing out any clashes that occur from picking and choosing glyphs in this manner.

Once this underlying system is in place, then it might be useful to build system of "styles", which would make a whole bunch of substitutions.  This could allow a font designer to create a style which systematically substitutes all the glyphs from their font into GregorioTeX, or allow a user to simply create and share a house style which combines already available glyphs in what ever manner they like.

The documentation would need a complete list of the glyph names, ideally in a table with images of how the glyph looks in each of our fonts, and possible alternatives.

**Comment by eroux** [comment]
I globally agree. I like the idea of having a big list as an annex of the documentation, this would indeed make life easier to users.

**Comment by henryso** [comment]
Just brainstorming here, but how about a macro like `\grealternative{targetName}[fontcsname]{nameOrNumber}`?  If `[fontcsname]` is left out, it will assume you want whatever `\gregoriofont` is and use the `\grecp` macro to fulfill a provided name.  If `[fontcsname]` is provided, the macro will resolve the name (or use the provided number) against that font.

**Comment by eroux** [comment]
I think that would be perfect yes (in LaTeX, the optional argument between square brackets is usually first, but it's not really important)

**Comment by henryso** [comment]
I did think about that, but that seems to convey the wrong message.

**Comment by eroux** [comment]
You mean the LaTeX syntax?

**Comment by henryso** [comment]
I mean that saying `\grealternative[font]{target}{replacement}` makes it seem like you are replacing `{target}` in `[font]` with `{replacement}` whereas what this macro is supposed to do is replace `{target}` with `{replacement}` from `[font]`.  I also considered `\grealternative[font]{replacement}{target}` but TeX is typically `\def\target{replacement}`, so I came up with that syntax (with the optional parameter in the middle).

**Comment by eroux** [comment]
Just a thought: maybe the font argument shouldn't be optional? All fons don't have to provide all alternatives, so an alternative is specific to a font...?

**Comment by henryso** [comment]
That's not a terrible idea.  I concur.

**Comment by henryso** [comment]
I think this is pretty much done, so I'm closing it, but I will open another issue for glyph documentation as mentioned by @rpspringuel.
